### ISSUE TITLE
Specify which classes in selector should receive variant prefix

### ISSUE THREAD

**Issue by adamwathan** [tailwindlabs/tailwindcss#3361]
It's come up a few times that folks want to be able to specify which classes in a selector are modified by variants (like `bg-black` becoming `md:bg-black`).

By default, Tailwind always targets the _last_ class in a selector, and that's because that's what we thought made the most sense for `group-hover`, which was the only real multi-class selector in Tailwind:

```css
/* Input */
@responsive {
  .group .group-hover\:bg-black {
    background-color: #000;
  }
}

/* Output */
.group .group-hover\:bg-black {
  background-color: #000;
}

@media (min-width: 640px) {
  .group .sm\:group-hover\:bg-black {
    background-color: #000;
  }
}

/* ... */
```

There have been a few situations where users have wanted different behavior, either applying the prefix to the first class, or to all classes.

Here is one such issue: https://github.com/tailwindlabs/tailwindcss/issues/2485

One solution that I think could work is using a special marker for the "variant target", perhaps using something like `[variant-target="my-class"]` (or even just `[.my-class]`, tested it with postcss-selector-parser and it works!) instead of `.my-class`. We would convert it to a real class at build time, but using something that looks like an attribute selector would be easily parseable by PostCSS, and work as a way to inform Tailwind that it should apply the variant to that part of the selector:

```css
/* Input */
@variants hover {
  [variant-target="foo"] .bar {
    /* ... */
  }
}

/* Output */
.hover\:foo:hover .bar {
  /* ... */
}

/* Input */
@variants hover {
  [variant-target="foo"] [variant-target="bar"] {
    /* ... */
  }
}

/* Output */
.hover\:foo:hover .hover\:bar:hover {
  /* ... */
}

/* Input */
@variants hover {
  .foo [variant-target="bar"] {
    /* ... */
  }
}

/* Output */
.foo .hover\:bar:hover {
  /* ... */
}

/* Input */
@variants hover {
  .foo .bar {
    /* ... */
  }
}

/* Output */
.foo .hover\:bar:hover {
  /* ... */
}
```

Not including this special syntax at all would preserve the existing behavior so there would be no backwards compatibility issues.

Opening this issue really just as a placeholder for when other issues that stem from this same problem pop up, so we can close them and point back to here. Any input on the problem or proposed solution would be great, even a PR to get things started would be cool as I don't have any concrete plans to prioritize this right away.

**Comment by gremo** [comment]
Hi @adamwathan , thanks for opening this issue that explain the problem in a better way. 

I like your proposed syntax, but what about making it a "parameter" for the `@responsive` or `@variant ` utilities? Like:

```css
@responsive 'strategy: all' {}
@variant hover, 'strategy: all' {}
```

I think it's more "explicit" about what are we are doing here. Where strategy could be `all`, `first`, `last` (default, as you said).  `strategy `could also be something like `prefix`, for say.

It would save a few keystrokes if one need to apply the utility to a lot of classes.

**Comment by saadeghi** [comment]
Hi, I'm having a similar problem trying to make a responsive component plugin.  
Here's an example: https://play.tailwindcss.com/Pp3V5L2HwA  
And the issue is `sm:` prefix applies to the last selector as you mentioned.  
I'm looking for a method to add it to the parent only.  

Plugin style:
```js
'.parent': {
  background: 'gray',
  '.child' : {
    background: 'yellow',
    '.grandchild' : {
      background: 'red',
    },
  },
}
```
Generated CSS is:
```css
@media (min-width: 640px){
  .sm\:parent {
    background: gray;
  }
  .parent .sm\:child {
    background: yellow;
  }
  .parent .child .sm\:grandchild {
    background: red;
  }
}
```
But I want it to be:
```css
@media (min-width: 640px){
  .sm\parent {
    background: gray;
  }
  .sm\parent .child {
    background: yellow;
  }
  .sm\:parent .child .grandchild{
    background: red;
  }
}
```

**Comment by fedetibaldo** [comment]
I'd like to chime in and say that it would be great to have such opportunity (to decide where the variants get applied) in custom plugins too. I looked around for a while, expecting a solution similar to the [manual selector prefix](https://tailwindcss.com/docs/plugins#manually-prefixing-selectors), but could find none.

This is how I'd expect it to work:
```js
module.exports = {
  variants: { custom: ['dark'] },
  plugins: [
    plugin(function({ addComponents, variant, variants }) {
      addComponents(
        {
          [`${variant('.new-class')} > .existing-class`]: {
            backgroundColor: '#fff',
          }
        },
        variants('custom')
      }
    })
  ]
}
```

**Comment by fedetibaldo** [comment]
So, I'm working on a PR, but it turns out that the variant generators aren't consistent in how they treat multi-class selectors. In particular:

- pseudo-class variants (e.g. `hover`, `focus`, `last-child`, etc.) add the variant prefix to every class in the selector.
```css
/* Input */
@variants hover {
  .banana .peel { color: yellow; }
}
/* Output */
.banana .peel { color: yellow; }
.hover\:banana:hover .hover\:peel:hover { color: yellow; }
```
- group variants (e.g. `group-hover`) intersperse the classes with an additional `.group:pseudo` class
```css
/* Input */
@variants group-hover {
  .banana .peel { color: yellow; }
}
/* Output */
.banana .peel { color: yellow; }
.group:hover .group-hover\:banana .group:hover .group-hover\:peel {
  color: yellow;
}
```

Only the responsive and motion variants (e.g. `motion-reduce`) use the last class as target, as we were all expecting. (I haven't tested the dark variant yet).

By logic, I assume the unexpected outputs to be incorrect. First, because some of them don't make any sense when transposed to their HTML form (e.g. one can't focus the parent AND the control element at the same time); second, because not a single test covers those cases, so if someone was relying on these weird outputs, they were definitely relying on a bug rather than a feature.

Hence, I'm writing test cases that expect all variants to use the last class as target by default (and fixing code along the way), but if anyone would like to provide me with some feedback and/or counter-arguments, please do. Otherwise, I'll just move on and sketch out the rest of the PR.

**Comment by chasegiunta** [comment]
@adamwathan is there currently a way to do this in custom plugins? If my variant applies a multi-class selector, can I specify which selector is prefixed with the responsive variant?

**Comment by adamwathan** [comment]
The only way to do it right now is to trick Tailwind by only using one real class, and using attribute selectors for the rest:

```css
[class~="poop"] .real-class [class~="pee"] {
  stupid: crap;
}
```

Then `.real-class` will get targeted but the others won't. Only downside to this is those attribute selectors will not have prefixes applied to them if you have the prefix option enabled in your config.

**Comment by RobinMalfait** [comment]
Hey all!

In v3 we improved the way how variants are handled. Instead of providing a custom syntax to mark the "target", we rely on the order when "using" the utilities and variants together. This means that we don't need any specific syntax when you define your utilities / variant implementations.

For people running into this issue, checkout: https://tailwindcss.com/docs/hover-focus-and-other-states#ordering-stacked-modifiers

The TL;DR is this:
```css
/* dark:group-hover:opacity-100 */
.dark .group:hover .dark\:group-hover\:opacity-100 {
  opacity: 1;
}

/* group-hover:dark:opacity-100 */
.group:hover .dark .group-hover\:dark\:opacity-100 {
  opacity: 1;
}
```

Here is a bigger visual example that explains it: https://play.tailwindcss.com/twRnDsWu1x?file=css
